### PR TITLE
feat: add SGLang RL metadata uploads via fsspec

### PR DESCRIPTION
**Pull request by Aphoh**
## Summary

Adds an opt-in SGLang metadata upload path for large RL-serving metadata such as routed expert captures, top-logprobs, tensors, and NumPy arrays. The feature is gated twice: the SGLang worker must run with `--enable-rl` / `DYN_SGL_ENABLE_RL=true`, and the request must include `nvext.metadata_upload.url`. Without RL mode, the request field is ignored.

The worker uploads one final zstd-compressed msgpack object when each choice finishes. Incremental chunks continue streaming normally. The final response is held until the upload succeeds; upload failure terminates the stream before the final response.

Uploads use the existing `dynamo.common.storage` fsspec wrapper, so the URL can target any installed fsspec backend.

## Request shape

```json
{
  "nvext": {
    "metadata_upload": {
      "url": "s3://bucket/root/rollouts/run-123"
    }
  }
}
```

The caller scopes each request by including its rollout/request identity in the URL. Objects are written as:

```text
<url>/choice_<index>.msgpack.zst
```

Unknown metadata upload fields, non-string URLs, and empty URLs are rejected.

## Payload

The decompressed msgpack object contains `schema_version` and the final SGLang `meta_info` under `metadata`. Torch tensors and NumPy arrays are normalized to `{type, dtype, shape, data}` with binary `data` preserved by msgpack.

Large metadata fields are omitted from streamed chunks when upload is enabled. Payload normalization and compression run off the asyncio event loop, and fsspec writes use the shared async storage helper.

## Validation

- 169 focused SGLang decode-handler and frontend tests passed
- Rust metadata upload schema test passed
- Rust backend extra-args passthrough test passed
- Black, isort, Ruff, flake8, codespell, and formatting checks passed
- `cargo fmt --all -- --check` passed
- `git diff --check` passed

### CODE REVIEW

File: components/src/dynamo/common/metadata_upload.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+def _serialize_zstd_json(payload: dict[str, Any]) -> bytes:
+    try:
+        import zstandard as zstd
+    except ImportError as exc:
+        raise RuntimeError(
+            "Metadata upload requires zstandard. "
+            "Install ai-dynamo with the selected backend extra or add the zstandard package."
+        ) from exc
+
+    raw = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode(
```
Comment (jthomson04) [inline components/src/dynamo/common/metadata_upload.py]: Is it possible to serialize with something other than json? The commas are just a bunch of redundant info, although zstd would probably take care of this. Or is this meant for maximizing compatibility?

Comment (Aphoh) [reply components/src/dynamo/common/metadata_upload.py]: Adding msgpack.

File: components/src/dynamo/sglang/request_handlers/llm/decode_handler.py [hunk 2]
```diff
@@ -62,13 +66,15 @@ def _extract_media_urls(mm_data: Dict[str, Any], media_key: str) -> list[str] |
...
 
 def _nvext_extra_field_requested(request: Dict[str, Any], field: str) -> bool:
-    nvext = request.get("nvext")
-    if not isinstance(nvext, dict):
-        return False
-    extra_fields = nvext.get("extra_fields")
-    if not isinstance(extra_fields, list):
-        return False
-    return field in extra_fields
+    extra_args = request.get("extra_args") or {}
```
Comment (dynamo-ops) [inline components/src/dynamo/sglang/request_handlers/llm/decode_handler.py:57]: `_nvext_extra_field_requested` now only reads `extra_args.nvext`, so existing backend callers that pass top-level `nvext.extra_fields` silently lose requested response fields like `stop_reason`. Fix: support both the new `extra_args.nvext.extra_fields` shape and the previous top-level `nvext.extra_fields` shape.

<details><summary>🤖 AI Fix</summary>

Modify `components/src/dynamo/sglang/request_handlers/llm/decode_handler.py::_nvext_extra_field_requested` to read `extra_fields` from both `request["extra_args"]["nvext"]` and `request["nvext"]` with dict/list type checks, and return true if `field` is present in either list.

</details>

File: components/src/dynamo/common/metadata_upload.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        )
+    else:
+        try:
+            import msgspec
+        except ImportError as exc:
+            raise RuntimeError("Metadata msgpack upload requires msgspec.") from exc
+        raw = msgspec.msgpack.encode(payload)
+
+    try:
+        return zstd.ZstdCompressor().compress(raw)
```
Comment (biswapanda) [inline components/src/dynamo/common/metadata_upload.py]: is this cpu bound? should we keep this configurable through payload_format as well
 { compression: zstd (default) , none } 

this can be a follow up

**Review by biswapanda — APPROVED** [review]
lgtm - few minor comments.

File: components/src/dynamo/common/metadata_upload.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+def _serialize_payload(payload: dict[str, Any], payload_format: str) -> bytes:
+    try:
+        import zstandard as zstd
+    except ImportError as exc:
+        raise RuntimeError(
+            "Metadata upload requires zstandard. "
+            "Install ai-dynamo with the selected backend extra or add the zstandard package."
+        ) from exc
+
+    if payload_format == "json":
```
Comment (biswapanda) [inline components/src/dynamo/common/metadata_upload.py]: Should we just keep `msgpack` and remove json ? 

```
import torch
meta_info = {"routed_experts": torch.arange(4, dtype=torch.float32).reshape(2, 2)}
mod._normalize_for_upload(meta_info)
mod._serialize_metadata(meta_info, "msgpack")
mod._serialize_metadata(meta_info, "json")
```

`_serialize_metadata(format='msgpack')` is able to serialize bytes properly
OK msgpack -> compressed len: 110

`_serialize_metadata(format='json')` fails 
json FAILED -> TypeError : Object of type bytes is not JSON serializable

Comment (Aphoh) [reply components/src/dynamo/common/metadata_upload.py]: Fixed this, removed json

File: components/src/dynamo/common/metadata_upload.py (hunk 3 above)
Comment (Aphoh) [reply components/src/dynamo/common/metadata_upload.py]: `asyncio.to_thread(_serialize_metadata, ...)` moves the whole normalization/encoding/compression pipeline off the event-loop thread. In `python-zstandard`’s C extension, the one-shot `compress()` call itself releases the GIL around zstd work, so this shouldn't block the event loop.

File: components/src/dynamo/sglang/request_handlers/llm/decode_handler.py (hunk 2 above)
Comment (Aphoh) [reply components/src/dynamo/sglang/request_handlers/llm/decode_handler.py:57]: This is already handled in the current branch: `_nvext_extra_field_requested` checks both top-level `nvext.extra_fields` and `extra_args.nvext.extra_fields`, with dict/list guards. `test_nvext_extra_field_requested_supports_raw_and_preprocessed_shapes` covers both forms.
